File: backend/src/main/java/com/backend/petplace/domain/review/dto/ReviewInfo.java
package com.backend.petplace.domain.review.dto;

import io.swagger.v3.oas.annotations.media.Schema;
import java.time.LocalDate;
import java.time.LocalDateTime;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Getter
@NoArgsConstructor
@Schema(description = "장소 리뷰 목록에 포함될 개별 리뷰 정보")
public class ReviewInfo {

  @Schema(description = "리뷰 ID", example = "10")
  private Long reviewId;

  @Schema(description = "작성자 이름", example = "멍멍이집사")
  private String userName;

  @Schema(description = "리뷰 내용", example = "정말 좋은 곳이었어요! 강아지가 좋아하네요.")
  private String content;

  @Schema(description = "별점", example = "5")
  private int rating;

  @Schema(description = "첨부 이미지 URL (없을 경우 null)", example = "https://s3.bucket/image.jpg")
  private String imageUrl;

  @Schema(description = "작성일", example = "2025-10-15")
  private LocalDate createdDate;

  public ReviewInfo(Long reviewId, String userName, String content, int rating, String imageUrl,
      LocalDateTime createdDate) {
    this.reviewId = reviewId;
    this.userName = userName;
    this.content = content;
    this.rating = rating;
    this.imageUrl = imageUrl;
    this.createdDate = createdDate.toLocalDate();
  }

  // S3 경로를 전체 URL로 교체하는 생성자
  private ReviewInfo(ReviewInfo dto, String fullImageUrl) {
    this.reviewId = dto.reviewId;
    this.userName = dto.userName;
    this.content = dto.content;
    this.rating = dto.rating;
    this.createdDate = dto.createdDate;
    this.imageUrl = fullImageUrl;
  }

  public static ReviewInfo withFullImageUrl(ReviewInfo dto, String fullImageUrl) {
    return new ReviewInfo(dto, fullImageUrl);
  }
}

File: backend/src/main/java/com/backend/petplace/domain/review/dto/ReviewInfo.kt
package com.backend.petplace.domain.review.dto

import io.swagger.v3.oas.annotations.media.Schema
import java.time.LocalDate
import java.time.LocalDateTime

@Schema(description = "장소 리뷰 목록에 포함될 개별 리뷰 정보")
data class ReviewInfo (
    @Schema(description = "리뷰 ID", example = "10")
    val reviewId: Long,

    @Schema(description = "작성자 이름", example = "멍멍이집사")
    val userName: String,

    @Schema(description = "리뷰 내용", example = "정말 좋은 곳이었어요! 강아지가 좋아하네요.")
    val content: String,

    @Schema(description = "별점", example = "5")
    val rating: Int,

    @Schema(description = "첨부 이미지 URL (없을 경우 null)", example = "https://s3.bucket/image.jpg")
    val imageUrl: String?,

    @Schema(description = "작성일", example = "2025-10-15")
    val createdDate: LocalDate
) {
    // JPQL Projection용 보조 생성자
    constructor(
        reviewId: Long,
        userName: String,
        content: String,
        rating: Int,
        imageUrl: String?,
        createdDate: LocalDateTime
    ) : this(
        reviewId = reviewId,
        userName = userName,
        content = content,
        rating = rating,
        imageUrl = imageUrl,
        createdDate = createdDate.toLocalDate()
    )

    companion object {
        // 이미지 URL 교체용 유틸리티 메서드 (copy 활용)
        // private constructor를 만들 필요 없이, copy() 메서드로 값만 바꿔서 새 객체 생성
        fun withFullImageUrl(dto: ReviewInfo, fullImageUrl: String?): ReviewInfo {
            return dto.copy(imageUrl = fullImageUrl)
        }
    }
}